### ISSUE TITLE
DropoutWrapper has unintended mask behavior when random seed is set

### ISSUE THREAD

**Issue by alquraishi** [tensorflow/tensorflow#2789]
When using the non-dynamic `rnn`, if the `seed` is set to something other than `None` for `DropoutWrapper`, the dropout masks for the inputs become synchronized across time steps. I.e. the same input dimension is dropped for all time steps for a given entry in a batch. If the maximum number of time steps varies between batches, then the later time steps begin to go out of sync (but all time steps that are earlier than the shortest sequence remain synchronized throughout). This behavior only affects the statically rolled out `rnn`. Dynamically rolled out RNNs using `dynamic_rnn` are always randomized per time step.

**Comment by ebrevdo** [comment]
Are you interested in contributing a fix?

**Comment by alquraishi** [comment]
@ebrevdo I'd be happy to, but it probably won't happen right away. I'm planning on expanding the suite of dropout functionality available (recurrence, zoneout, etc) and can include it then. But it would likely be several weeks after ICML.

**Comment by ebrevdo** [comment]
Wonderful!  For now I'll mark this as "contributions welcome", and feel free to contribute this fix into core and/or any other types of dropout (citing your work, of course), if you like, in tf.contrib.rnn.

**Comment by LeavesBreathe** [comment]
@alquraishi what you are suggesting only applies when you use the dropout wrapper from `rnn_cell.py`. However, if you just apply dropout at the `rnn.py` level, you can easily adjust the stochastic dropout masks should change on each timestep. 

I'm working on implementing zoneout right now and I think that all you need to do is use `tf.nn.dropout` to achieve this with `rnn.py`

**Comment by alquraishi** [comment]
@LeavesBreathe yes I'm aware but I wanted to keep the clean software design of the dropout wrapper. Btw, I have already implemented zoneout, and will be submitting a PR.

**Comment by LeavesBreathe** [comment]
Thanks @alquraishi . On your implementation of zoneout, did you multiply the outputs by `1 - (1/d)` in order to maintain the sum of the outputs the same? I tried to see if they did that in the paper and couldn't find if they did that or not.

Also, did you end up using the LSTM variant that they showed right before section 4? http://arxiv.org/pdf/1606.01305v2.pdf

I find that these elements may be crucial and could explain why they used a d of 0.05 (which is pretty low) for optimal results.

**Comment by teganmaharaj** [comment]
@alquraishi @LeavesBreathe 
Multiplying by 1-1/d (and then doing the appropriate test time correction) is definitely crucial; in the version on arxiv we had not yet implemented zoneout in tensorflow so we didn't mention anything about it. 
I'm happy to help with this if it's not done already? (I'm one of the zoneout authors)

**Comment by davek44** [comment]
I'd be a zoneout consumer if one of these implementations is ready.

**Comment by wt-huang** [comment]
Closing as this is resolved
